extensions/amp-ad-network-adzerk-impl/0.1/amp-ad-network-adzerk-impl.js

```diff
@@ -0,0 +1,175 @@
+
+/** @typedef {{
+      templateUrl: string,
+      data: (JsonObject|undefined),
+    }} */
+let AmpTemplateCreativeDef;
+
+/** @private {?AmpAdTemplates} */
+let ampAdTemplates;
+
+/**
+ * Fast Fetch implementation for AdZerk network that allows AMP creative
+ * preferential render via AMP cache stored template expansion using
+ * amp-mustache.  AMP creative response will consist of the following JSON
+ * object with two fields:
+ *
+ * - templateUrl: number value for template ID.  Template must already
+ *    have been stored in the AMP cache.
+ * - data: optional JSON object mapping of macro name to its
+ *    string value used to dynamically update the template
+ *
+ * Additionally, ad response must include header indicating AMP creative
+ * template type, e.g.: `AMP-template-amp-creative: amp-mustache`.
+ *
+ * Failure to properly fetch or expand template will result in slot collapsing.
+ * Non-AMP creatives (defined as those not including AMP-template-amp-creative)
+ * will be rendered via cross domain frame.
+ */
+export class AmpAdNetworkAdzerkImpl extends AmpA4A {
+
+  /**
+   * @param {!Element} element
+   */
+  constructor(element) {
+    super(element);
+
+    /** @private {?CreativeMetaDataDef} */
+    this.creativeMetadata_ = null;
+
+    /** @private {?AmpTemplateCreativeDef} */
+    this.ampCreativeJson_ = null;
+
+    ampAdTemplates = ampAdTemplates || new AmpAdTemplates(this.win);
+  }
+
+  /**
+   * Validate the tag parameters.  If invalid, ad ad will not be displayed.
+   * @override
+   */
+  isValidElement() {
+    return !!this.getAdUrl();
+  }
+
+  /** @override */
+  getSigningServiceNames() {
+    // Does not utilize crypto signature based AMP creative validation.
+    // TODO(keithwrightbos): move import of crypto validation into
+    // implementations, reducing adzerk binary size.
+    return [];
+  }
+
+  /** @override */
+  getAdUrl() {
+    const src = this.element.getAttribute('src');
+    if (!/^https:\/\/adzerk.com\?id=\d+$/i.test(src)) {
+      return '';
+    }
+    if (getMode(this.win).localDev) {
+      return `http://ads.localhost:${this.win.location.port}` +
+        '/adzerk/' + /^https:\/\/adzerk.com\?id=(\d+)/.exec(src)[1];
+    }
+    // TODO(adzerk): specify expected src path.
+    return /^https:\/\/adzerk.com\?id=\d+$/i.test(src) ? src : '';
+  }
+
+  /** @override */
+  maybeValidateAmpCreative(bytes, headers) {
+    if (headers.get(AMP_TEMPLATED_CREATIVE_HEADER_NAME) !== 'true') {
+      return /**@type {!Promise<(ArrayBuffer|null)>}*/ (Promise.resolve(null));
+    }
+    // Shorthand for: reject promise if current promise chain is out of date.
+    const checkStillCurrent = this.verifyStillCurrent();
+    return utf8Decode(bytes).then(body => {
+      checkStillCurrent();
+      this.ampCreativeJson_ = /** @type {!AmpTemplateCreativeDef} */
+        (tryParseJson(body) || {});
+      const proxyUrl = getMode(this.win).localDev
```

Comment:
can you move the proxy URL logic into AmpAdTemplates for sharing?

Reply:
Done.